### PR TITLE
Add intl4x number format

### PR DESCRIPTION
**Pull request by mosuem**
Adding a number formatting API. This includes the stub for ICU4X as well as the working browser implementation.

For tests, I am currently looking for a good way to get the intl402 part of the [test262](https://github.com/tc39/test262) suite transferred to a format where I could consume it here.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

#### Contribution guidelines:

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Many Dart repos have a weekly cadence for reviewing PRs - please allow for a week or two of latency for initial review feedback.

---

### CODE REVIEW

**Review by devoncarew — APPROVED** [review]
In this PR or the next one, we should probably look to expand either the sample, the tests, or both, as a way to get a sense of what the API is like to use.

File: pkgs/intl4x/lib/src/number_format/number_format_options.dart [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+enum SignDisplay {
+  auto,
+  always,
+  exceptZero,
+  negative,
+  never;
+}
+
+enum Unit {
```
[the remark this answers is not in the document]
Comment (mosuem) [reply pkgs/intl4x/lib/src/number_format/number_format_options.dart:144]: Yes, I am currently in discussions with legal about this.

File: pkgs/intl4x/lib/src/number_format/number_format.dart [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+import '../ecma/ecma_policy.dart';
+import '../locale.dart';
+import '../options.dart';
+import '../test_checker.dart';
+import '../utils.dart';
+import 'number_format_4x.dart';
+import 'number_format_options.dart';
+import 'number_format_stub.dart' if (dart.library.js) 'number_format_ecma.dart';
+
+/// Number formatting functionality of the browser.
```
[the remark this answers is not in the document]
Comment (mosuem) [reply pkgs/intl4x/lib/src/number_format/number_format.dart:14]: I wasn't sure where to best do the conditional switch.

**Comment by mosuem** [comment]
Regarding the tests, the plan is to integrate with the [conformance project](https://github.com/unicode-org/conformance/) from unicode.

**Review by robertbastian — COMMENTED** [review]
looks good

**Comment by mosuem** [comment]
@Reprevise Just pending some legal questions, hopefully in the next days!

**Pull request merged**